src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs

```diff
@@ -165,18 +166,104 @@ public class PipelineArtifactServer
+                }
+            }
+            else if (downloadOptions == DownloadOptions.SingleDownload)
+            {
+                // 1) get manifest id from artifact data
+                BuildArtifact buildArtifact;
+                if (downloadParameters.ProjectRetrievalOptions == BuildArtifactRetrievalOptions.RetrieveByProjectId)
+                {
+                    buildArtifact = await buildHelper.GetArtifact(downloadParameters.ProjectId, downloadParameters.PipelineId, downloadParameters.ArtifactName, cancellationToken);
+                }
+                else if (downloadParameters.ProjectRetrievalOptions == BuildArtifactRetrievalOptions.RetrieveByProjectName)
+                {
+                    if (string.IsNullOrEmpty(downloadParameters.ProjectName))
+                    {
+                        throw new InvalidOperationException("Project name can't be empty when trying to fetch build artifacts!");
+                    }
+                    else
+                    {
+                        buildArtifact = await buildHelper.GetArtifactWithProjectNameAsync(downloadParameters.ProjectName, downloadParameters.PipelineId, downloadParameters.ArtifactName, cancellationToken);
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException("Unreachable code!");
+                }
+                ArtifactProviderFactory factory = new ArtifactProviderFactory(context, connection, this.CreateTracer(context));
+                IArtifactProvider provider = factory.GetProvider(buildArtifact);
+                await provider.DownloadSingleArtifactAsync(downloadParameters, buildArtifact, cancellationToken);
+            }
+            else
+            {
+                throw new InvalidOperationException("Unreachable code!");
+            }
+        }
         private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext context, VssConnection connection)
         {
             var dedupStoreHttpClient = connection.GetClient<DedupStoreHttpClient>();
-            var tracer = new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information);
+            var tracer = this.CreateTracer(context);
             dedupStoreHttpClient.SetTracer(tracer);
             var client = new DedupStoreClientWithDataport(dedupStoreHttpClient, 16 * Environment.ProcessorCount);
             var buildDropManager = new BuildDropManager(client, tracer);
             return buildDropManager;
         }
+
+        private CallbackAppTraceSource CreateTracer(AgentTaskPluginExecutionContext context)
+        {
+            var tracer = new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information);
+            return tracer;
+        }
     }
 
-    internal class PipelineArtifactDownloadParameters
+    public class PipelineArtifactDownloadParameters
```

Comment:
>public [](start = 4, length = 6)

why are we changing these visbilities

Reply:
Need this to be public for the IArtifactProvider.

Reply:
IArtifactProvider can be internal as well so can the factory, this can be plumbed through

---
In reply to: [283576154](https://github.com/microsoft/azure-pipelines-agent/pull/2210#discussion_r283576154) [](ancestors = 283576154)